### PR TITLE
Post "Distributed Training with PySpark".

### PR DESCRIPTION
**Pull request by vhdeluca**
Please, review the new post "Distributed Training with PySpark".

### CODE REVIEW

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+categories:
+  - data-science
+tags:
+  - distributed-training
+  - pyspark
+  - ensemble
+excerpt: ""
+---
+
+There are several definitions around the term "Big Data" all over the internet, books and so on. In practical terms, a data scientist "feels" a Big Data scenario every time the amount of data does not fit in memory. From this point on, there are several software available on the internet that try to solve this scenario, but one of the most common ways would be to work with PySpark.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:12]:
```suggestion
There are several definitions around the term "Big Data" all over the internet, books, and so on. In practical terms, a data scientist "feels" a Big Data scenario every time the amount of data does not fit in memory. From this point on, there are several tools available on the internet that try to solve this problem, but one of the most common ways would be to work with PySpark.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+
+Once PySpark is in use and the data is distributed across a cluster, various functions can be used to transform the data. However, the strategy on how to train your AI model may not yet be defined. Several options are possible. Let's try to list them:
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:14]:
```suggestion
Once PySpark is in use and the data is distributed across a cluster, various functions can be used to transform the data. However, the strategy on how to train your AI model may not yet be defined. Several options are available. Let's try to list them:
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+
+
+# Seaborn config
+sns.set(style="darkgrid")
+sns.set(rc={'figure.figsize':(17,5)})
+```
+
+## Defining the synthetic dataset
+
+For the purpose of this experiment, we are going to use a synthetic dataset using sklearn datasets.make_classification to generate the data. The dataset will contain one million instances and 20 features. This number could be even higher for the distributed approach, but we need to compare it to the centralized approach and therefore we cannot increase it that much.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:55]:
```suggestion
For this experiment, we are going to use a synthetic dataset using sklearn `datasets.make_classification` to generate the data. The dataset will contain one million instances and 20 features. This number could be even higher for the distributed approach, but we need to compare it to the centralized approach and therefore we cannot increase it that much.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+    distributed_preds = distributed_preds + preds/N_FOLDS
+    
+accuracy_score(y_test, np.rint(distributed_preds))
+```
+
+    0.9043466666666666
+
+Accuracy reached 0.9043. Note that this accuracy is very close to the accuracy obtained with the first centralized approach (0.9045). The difference is in the execution time: distributed training takes longer (21:21 minutes), approximately 5.5 times longer for this case shown. 
+
+Also, we did the following experiment: we ran this experiment for 29 different dataset sizes (e.g. 500k rows, 100k rows, etc) and the RMSE between centralized and distributed approaches was 0.00234! That is, a very small number compared to the accuracies obtained.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:224]:
```suggestion
Also, we did the following experiment: we ran this experiment for 29 different dataset sizes (e.g. 500k rows, 100k rows, etc) and the RMSE between centralized and distributed approaches was 0.00234! This is a very small number compared to the accuracies obtained. 
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+    upload_file()
+
+    # 6. Return results as pandas DF
+    res = pd.DataFrame({'fold': fold,
+                        'Accuracy': accuracy,
+                        'model_path': model_path}, index=[0])
+    return res
+```
+
+In this part, a groupby is performed on the "fold" column and a command is applied to send the Pandas UDF function "run_model_by_fold" to the nodes. Note that this Spark command is lazy.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:297]:
```suggestion
In this part, a groupby is performed on the *fold* column and a command is applied to send the Pandas UDF function `run_model_by_fold` to the nodes. Note that this Spark command is lazy.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+```python
+from sklearn.ensemble import RandomForestClassifier
+
+clf = RandomForestClassifier(max_depth=3, random_state=123)
+clf.fit(X_train, y_train)
+
+ensemble_preds = clf.predict(X_test)
+```
+
+The final accuracy is 0.8965, an intermediate result between the ensemble with the mean and the ensemble with the median. Overall, a great result. As the difference is small in relation to the ensemble with the mean, it is not possible to say which one is the best. Also because another model could have been used (instead of Random Forest) and even the RF could be tested with other hyperparameters. In our experiences, in some results the stacking was better than the ensemble with the average.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:466]:
```suggestion
The final accuracy is 0.8965, an intermediate result between the ensemble with the mean and the ensemble with the median. Overall, a great result. As the difference is small in relation to the ensemble with the mean, it is not possible to say which one is the best. Also because another model could have been used (instead of Random Forest) and even the same model could be tested with other hyperparameters. In our experiences, in some results, the stacking was better than the ensemble with the average.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+
+It is important to remember that the accuracy of the third and fourth approaches tend to have a greater difference from the accuracy of the initial centralized approach when the dataset size decreases. But if that happens, the centralized training shown in the first approach becomes a real possibility.
+
+Finally, there is no approach considered the best or ideal. It all depends on your scenario.
+
+We hope you enjoyed the content. Let us know if you have any questions or suggestions.
+
+
+*Authors: [Vitor Hugo Medeiros De Luca](https://www.linkedin.com/in/vitordeluca/), [Igor Siqueira Cortez](https://www.linkedin.com/in/igor-cortez-56793825/), [Luiz Felipe Manke](https://br.linkedin.com/in/luizmanke/)*
+{: .notice}
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:495]: What is this?

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+There are several definitions around the term "Big Data" all over the internet, books and so on. In practical terms, a data scientist "feels" a Big Data scenario every time the amount of data does not fit in memory. From this point on, there are several software available on the internet that try to solve this scenario, but one of the most common ways would be to work with PySpark.
+
+Once PySpark is in use and the data is distributed across a cluster, various functions can be used to transform the data. However, the strategy on how to train your AI model may not yet be defined. Several options are possible. Let's try to list them:
+
+1. Centralize all data to train the model;
+2. Train in a fully distributed way;
+3. Centralize a representative sample of the data;
+4. Train on multiple samples and then ensemble them.
+
+Option 1 will be discussed only for the sake of benchmarking, as it would not be possible by the definition of the problem presented by this post. Options 2, 3 and 4 will be shown with their pros and cons.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:21]:
```suggestion
Option 1 will be discussed only for the sake of benchmarking, as it would not be possible by the definition of the problem presented by this post. Options 2, 3, and 4 will be shown with their pros and cons.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+# Seaborn config
+sns.set(style="darkgrid")
+sns.set(rc={'figure.figsize':(17,5)})
+```
+
+## Defining the synthetic dataset
+
+For the purpose of this experiment, we are going to use a synthetic dataset using sklearn datasets.make_classification to generate the data. The dataset will contain one million instances and 20 features. This number could be even higher for the distributed approach, but we need to compare it to the centralized approach and therefore we cannot increase it that much.
+
+In order to make things a bit harder, we are going to configure only 10 features as informative, flip 10% of the target values and short the distance between classes. Why this? Because we choosed this way. You are welcome to reproduce this experiment with other combinations. Have fun!
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:57]:
```suggestion
In order to make things a bit harder, we are going to configure only 10 features as informative, flip 10% of the target values and shorten the distance between classes. Why this? Because we chose this way. You are welcome to reproduce this experiment with other combinations. Have fun!
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+N_FOLDS = 5
+
+skf = KFold(n_splits=N_FOLDS, shuffle=True, random_state=123)
+
+folds_index = list(skf.split(X_train, y_train))
+```
+
+## First approach: centralized training
+
+In the first approach, we are going to use a centralized CatBoost model with the default hyperparameters set. The metric we are going to use to compare the approachs is the accuracy since the target is balanced and we do not have any business considerations here.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:94]:
```suggestion
In the first approach, we are going to use a centralized CatBoost model with the default hyperparameters set. The metric we are going to use to compare the approaches is the accuracy since the target is balanced and we do not have any business considerations here.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+
+One model is expected for each fold reaching a total of 5 models. For each model, 4 folds are used for training and 1 fold used for validation. This is shown below.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:96]:
```suggestion
One model is expected for each fold reaching a total of 5 models. For each model, 4 folds are used for training, and 1 fold is used for validation. This is shown below.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+accuracy_score(y_test, np.rint(centralized_preds))
+```
+
+    0.9045
+
+
+
+Accuracy reached 0.9045. This number alone doesn't say anything because we're interested in making comparisons with upcoming approaches. So it is a benchmark number. 
+
+Before continuing reading, try to think about what the expected results would be and only after continuing reading.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:138]:
```suggestion
Before continuing reading, try to think about what the expected results would be, and only after continuing reading.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+
+## Second approach: distributed training
+
+In the second approach, we are going to use a distributed CatBoost model with the default hyperparameters set. The same cross-validation is performed as done in the first approach, but the data always keeps distributed across the entire cluster.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:142]:
```suggestion
In the second approach, we are going to use a distributed CatBoost model with the default hyperparameters set. The same cross-validation is performed as done in the first approach, but the data always remains distributed across the entire cluster.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 14]
```diff
@@ -0,0 +1,495 @@
...
+    # Convert to Pool format
+    trainPool = PoolSpark(dfs_train_fit)
+    evalPool = PoolSpark(dfs_valid_fit)
+    
+    classifier_fitted.append(spark_models[i].fit(trainPool, [evalPool]))
+```
+
+    5it [21:21, 256.21s/it] 
+
+As with the first centralized approach, we have 5 trained models and need to perform the predictions on the test set. An average among all models shows the final prediction, but in this case we need to perform all the conversions, just like we did in the training step.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:194]:
```suggestion
As with the first centralized approach, we have 5 trained models and need to perform the predictions on the test set. An average among all models shows the final prediction, but in this case, we need to perform all the conversions, just like we did in the training step.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 15]
```diff
@@ -0,0 +1,495 @@
...
+
+# Insert 'fold' column
+for i, (train_index, valid_index) in enumerate(folds_index):
+    df_train.loc[valid_index, 'fold'] = int(i)
+    
+# Convert to Spark DF
+dfs_replication = spark.createDataFrame(df_train)
+```
+
+In the code below, Pandas UDF is defined. This function will be sent to each node in order to perform the training. If you are struggling to understand how Pandas UDF works, please read the post "[Leveraging Machine Learning Tasks with PySpark Pandas UDF](https://neowaylabs.github.io/data-science/Leveraging-Machine-Learning-Tasks-with-PySpark-Pandas-UDF/)".
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:247]:
```suggestion
In the code below, Pandas UDF is defined. This function will be sent to each node to perform the training. If you are struggling to understand how Pandas UDF works, please read the post "[Leveraging Machine Learning Tasks with PySpark Pandas UDF](https://neowaylabs.github.io/data-science/Leveraging-Machine-Learning-Tasks-with-PySpark-Pandas-UDF/)".
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 16]
```diff
@@ -0,0 +1,495 @@
...
+    # Accuracy
+    print(f"Model {i}'s accuracy: " + str(accuracy_score(df_test['label'].values, df_test[f'preds_model_{i}'].apply(np.rint).values)))
+```
+    Model 0's accuracy: 0.8922566666666667
+    Model 1's accuracy: 0.8925666666666666
+    Model 2's accuracy: 0.89263
+    Model 3's accuracy: 0.89216
+    Model 4's accuracy: 0.89311
+
+Now we have the accuracy calculated across the test set. Note that they are very similar between each other and only fluctuate on the third decimal place. Note also that, despite coming from models trained on a sample, these accuracies are very close to the approaches that train on the entire dataset.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:346]:
```suggestion
Now we have the accuracy calculated across the test set. Note that they are very similar to each other and only fluctuate in the third decimal place. Note also that, despite coming from models trained on a sample, these accuracies are very close to the approaches that train on the entire dataset.
```

File: _posts/2022-01-20-distributed-training-with-pyspark.md [hunk 17]
```diff
@@ -0,0 +1,495 @@
...
+
+The first centralized approach achieved an accuracy of 0.9045 and a time of 46.36s per fold. However, this approach is impractical when the amount of data is really large and was used here only as a benchmark for the other approaches.
+
+The second approach achieved an accuracy of 0.9043 and a time of 256.21s per fold. That is, an accuracy practically equal to the centralized approach with a considerably longer training and prediction time given the decentralized nature of the solution.
+
+The third approach trained several models centralized on different samples of the dataset reaching accuracies ranging from 0.89216 to 0.89311 in a time of 26.53s. That is, an accuracy close to the initial centralized approach with a shorter training and prediction time. It is important to note that the training was centralized, but was carried out in the nodes in a distributed way, reducing the time required for training.
+
+The fourth approach showed how the third approach can come even closer to the first centralized approach. Through the ensemble of the different trained models it is possible to achieve even greater accuracies.
```
Comment (luizmanke) [inline _posts/2022-01-20-distributed-training-with-pyspark.md:485]:
```suggestion
The fourth approach showed how the third approach can come even closer to the first centralized approach. Through the ensemble of the different trained models, it is possible to achieve even greater accuracies.
```

**Review by luizmanke — APPROVED** [review]
Just minor text suggestions

File: _posts/2022-01-20-distributed-training-with-pyspark.md (hunk 7 above)
Comment (vhdeluca) [reply _posts/2022-01-20-distributed-training-with-pyspark.md:495]: It is necessary to leave a gray rectangle behind the authors.
